### PR TITLE
monasca: Add openvswitch plugin (SCRD-7571)

### PR DESCRIPTION
**Pull request by witekest**
Add Monasca OVS plugin configuration on all neutron-network nodes.
Plugin configuration is automatically enabled and can be disabled in
Crowbar UI.

Monasca OVS plugin autodetection is disabled on Monasca server node.

Add `monasca-agent` role to `delegate_authorized_roles` to enable
collecting tenant metrics.

### CODE REVIEW

File: chef/cookbooks/monasca/templates/default/monasca-api.conf.erb [hunk 1]
```diff
@@ -624,6 +624,7 @@ agent_authorized_roles = monasca-agent
 # behalf of another tenant
 #  (list value)
 #delegate_authorized_roles = admin
```
Comment (cmurphy) [inline chef/cookbooks/monasca/templates/default/monasca-api.conf.erb]: delete this?

File: chef/data_bags/crowbar/migrate/monasca/315_add_ovs_plugin_settings.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+def upgrade(template_attrs, template_deployment, attrs, deployment)
+  attrs["agent"]["monitor_ovs"] = template_attrs["agent"]["monitor_ovs"] unless
+    attrs["agent"].key?("monitor_ovs")
+
+  attrs["agent"]["plugins"] = template_attrs["agent"]["plugins"] unless
+    attrs["agent"].key?("plugins")
```
Comment (cmurphy) [inline chef/data_bags/crowbar/migrate/monasca/315_add_ovs_plugin_settings.rb]: I could be wrong but I don't think this will do what you want, the template already has a pre-existing "plugins" dictionary, so either this will not add in the new "ovs" key or it will overwrite the old version with the defaults in the template instead of the user-set values. I think this should be `attrs["agent"]["plugins"]["ovs"] = template_attrs... etc`

File: chef/cookbooks/monasca/templates/default/monasca-api.conf.erb (hunk 1 above)
Comment (witekest) [reply chef/cookbooks/monasca/templates/default/monasca-api.conf.erb]: Done

File: chef/data_bags/crowbar/migrate/monasca/315_add_ovs_plugin_settings.rb (hunk 2 above)
Comment (witekest) [reply chef/data_bags/crowbar/migrate/monasca/315_add_ovs_plugin_settings.rb]: Good catch. Thank you.

File: chef/cookbooks/monasca/providers/agent_plugin_ovs.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+action :create do
+  config = {
+    "username" => new_resource.username,
+    "user_domain_name" => new_resource.user_domain_name,
+    "password" => new_resource.password,
+    "project_name" => new_resource.project_name,
+    "project_domain_name" => new_resource.project_domain_name,
+    "auth_url" => new_resource.auth_url,
+    "region_name" => new_resource.region_name
+  }.merge(node[:monasca][:agent][:plugins][:ovs])
```
Comment (toabctl) [inline chef/cookbooks/monasca/providers/agent_plugin_ovs.rb]: in the other providers, we use `node[:monasca][:agent_plugin_config][:ovs_instances]` to store the different configs. IIRC the idea was, that a use could also store custom configs in the node object if needed.
Is there a reason to do this differently here?

File: chef/cookbooks/monasca/providers/agent_plugin_ovs.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  # be sure the package is installed. that way "/etc/monasca/agent/conf.d/" is available
+  # and also the user and group are there
+  package "openstack-monasca-agent"
+
+  # NOTE(toabctl): convert/parse first to/from json. Otherwise we have unwanted markers
+  # like "- !ruby/hash:Mash" in the yaml output
+  process_conf = JSON.parse({ "init_config" => config,
```
Comment (toabctl) [inline chef/cookbooks/monasca/providers/agent_plugin_ovs.rb]: nit: ovs_conf would be a better name

File: chef/cookbooks/monasca/templates/default/monasca-api.conf.erb [hunk 5]
```diff
@@ -623,7 +623,7 @@ agent_authorized_roles = monasca-agent
 # Roles that are allowed to POST metrics on
 # behalf of another tenant
 #  (list value)
-#delegate_authorized_roles = admin
+delegate_authorized_roles = monasca-agent
```
Comment (toabctl) [inline chef/cookbooks/monasca/templates/default/monasca-api.conf.erb]: This is unrelated to the ovs agent enablement? if so, can you move it to another commit please?

File: chef/cookbooks/neutron/recipes/monitor_monasca_ovs.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#
```
Comment (toabctl) [inline chef/cookbooks/neutron/recipes/monitor_monasca_ovs.rb]: Shouldn't that be in the file `monitor_monasca.rb` which already contains the http_check configuration?

File: chef/cookbooks/neutron/recipes/monitor_monasca_ovs.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+return unless node["roles"].include?("monasca-agent")
+
+keystone_settings = KeystoneHelper.keystone_settings(node, @cookbook_name)
+
+if node[:monasca][:agent][:monitor_ovs]
```
Comment (toabctl) [inline chef/cookbooks/neutron/recipes/monitor_monasca_ovs.rb]: Is there a reason to **not** monitor OVS if monasca is deployed? So why do we need to make it configurable?

File: chef/cookbooks/monasca/providers/agent_plugin_ovs.rb (hunk 3 above)
Comment (witekest) [reply chef/cookbooks/monasca/providers/agent_plugin_ovs.rb]: Instances section of OVS plugin configuration [should be empty](https://github.com/openstack/monasca-agent/blob/master/docs/Ovs.md).

File: chef/cookbooks/monasca/providers/agent_plugin_ovs.rb (hunk 4 above)
Comment (witekest) [reply chef/cookbooks/monasca/providers/agent_plugin_ovs.rb]: Done.

File: chef/cookbooks/monasca/templates/default/monasca-api.conf.erb (hunk 5 above)
Comment (witekest) [reply chef/cookbooks/monasca/templates/default/monasca-api.conf.erb]: Done.

File: chef/cookbooks/neutron/recipes/monitor_monasca_ovs.rb (hunk 6 above)
Comment (witekest) [reply chef/cookbooks/neutron/recipes/monitor_monasca_ovs.rb]: http_check from `monasca_monasca.rb` should be configured on Neutron server.
OVS plugin should be configured on all Neutron agents.
The recipes are included in `role_neutorn_server.rb` and `role_neutron_network_network.rb` respectively.

File: chef/cookbooks/neutron/recipes/monitor_monasca_ovs.rb (hunk 7 above)
Comment (witekest) [reply chef/cookbooks/neutron/recipes/monitor_monasca_ovs.rb]: I have just followed the same pattern as for libvirt and Ceph plugins. OVS plugin was not supported till now, so adding the configuration option increases visibility for end users, who not necessarily read release notes. I don't have a strong opinion though.

File: chef/data_bags/crowbar/template-monasca.json [hunk 8]
```diff
@@ -32,6 +32,18 @@
             "vm_ping_check_enable": true,
             "vm_probation": 300,
             "vnic_collection_period": 0
+          },
+          "ovs": {
+            "cache_dir": "/dev/shm",
+            "neutron_refresh": 14400,
+            "network_use_bits": false,
+            "check_router_ha": true,
```
Comment (toabctl) [inline chef/data_bags/crowbar/template-monasca.json]: `check_router_ha` should be `false` if no router HA is used. Documentation says:

_check_router_ha will check router HA status if set to true. This should be set to false if not configuring routers for HA, as setting this to true will cause the plugin to make additional neutron calls._

I think you need to check for `neutron[:neutron][:l3_ha][:use_l3_ha]`

Given that this depends on `neutron[:neutron][:l3_ha][:use_l3_ha]`, I would remove the config option and just set it correctly depending on `use_l3_ha`

File: chef/data_bags/crowbar/template-monasca.json [hunk 9]
```diff
@@ -32,6 +32,18 @@
             "vm_ping_check_enable": true,
             "vm_probation": 300,
             "vnic_collection_period": 0
+          },
+          "ovs": {
+            "cache_dir": "/dev/shm",
```
Comment (toabctl) [inline chef/data_bags/crowbar/template-monasca.json]: do you know how much cached data will be produced? on a mkcloud deployment, I have 6 GB free on `/dev/shm`

Comment (witekest) [reply chef/data_bags/crowbar/template-monasca.json]: This includes only values from the last Neutron API call. In my test env I measured ~1KB per router/port. I don't expect any issues here.

File: chef/data_bags/crowbar/template-monasca.json (hunk 8 above)
Comment (witekest) [reply chef/data_bags/crowbar/template-monasca.json]: Good point. Done.
